### PR TITLE
planner: regard NULL as point when accessing composite index

### LINKED ISSUE #29650: planner: NullEQ cannot be used for composite indexes
## Enhancement
```
mysql> show create table t;
+-------+------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                               |
+-------+------------------------------------------------------------------------------------------------------------------------------------------------------------+
| t     | CREATE TABLE `t` (
  `a` int(11) DEFAULT NULL,
  `b` int(11) DEFAULT NULL,
  KEY `a` (`a`,`b`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin |
+-------+------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> explain select * from t where a<=>null and b=1;
+--------------------------+---------+-----------+------------------------+---------------------------------------------------+
| id                       | estRows | task      | access object          | operator info                                     |
+--------------------------+---------+-----------+------------------------+---------------------------------------------------+
| IndexReader_7            | 0.01    | root      |                        | index:Selection_6                                 |
| └─Selection_6            | 0.01    | cop[tikv] |                        | eq(test.t.b, 1)                                   |
|   └─IndexRangeScan_5     | 10.00   | cop[tikv] | table:t, index:a(a, b) | range:[NULL,NULL], keep order:false, stats:pseudo |
+--------------------------+---------+-----------+------------------------+---------------------------------------------------+
3 rows in set (0.00 sec)
```

The range should be `[NULL 1, NULL 1]`.

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve?

Issue Number: close #29650

Problem Summary: planner: regard NULL as point when accessing composite index

### What is changed and how it works?

planner: regard NULL as point when accessing composite index

### Check List

Tests 

- [x] Unit test

### Release note

```release-note
planner: regard NULL as point when accessing composite index
```

### CODE REVIEW

File: util/ranger/ranger_test.go [hunk 1]
```diff
@@ -961,6 +961,7 @@ func TestCompIndexDNFMatch(t *testing.T) {
 	require.NoError(t, err)
 	testKit := testkit.NewTestKit(t, store)
 	testKit.MustExec("use test")
+	testKit.MustExec(`set @@session.tidb_regard_null_as_point=false`)
```
Comment (time-and-fate) [inline util/ranger/ranger_test.go:964]: What would happen if we don't add this line.

File: distsql/request_builder.go [hunk 2]
```diff
@@ -715,6 +715,8 @@ func encodeIndexKey(sc *stmtctx.StatementContext, ran *ranger.Range) ([]byte, []
 		}
 	}
 
+	// NOTE: this is a hard-code operation to avoid wrong results when accessing unique index with NULL;
+	// Please see https://github.com/pingcap/tidb/issues/29650 for more details
```
Comment (xuyifangreeneyes) [inline distsql/request_builder.go:719]: Will we remove the hard-code operation in the following PRs?

File: planner/core/plan_to_pb.go [hunk 3]
```diff
@@ -212,6 +212,22 @@ func checkCoverIndex(idx *model.IndexInfo, ranges []*ranger.Range) bool {
 		if len(rg.LowVal) != len(idx.Columns) {
 			return false
 		}
+		if !rg.LowExclude {
```
Comment (xuyifangreeneyes) [inline planner/core/plan_to_pb.go:215]: Do we need to check whether null exists if `rg.LowExclude or rg.HighExclude`?

Comment (xuyifangreeneyes) [inline planner/core/plan_to_pb.go:215]: Do we need to check whether null exists if `rg.LowExclude or rg.HighExclude`?

**Comment by qw4990** [comment]
More test cases will be covered by https://github.com/pingcap/qa/blob/master/comp/yy/nullable_index.yy.

File: util/ranger/ranger_test.go (hunk 1 above)
Comment (qw4990) [reply util/ranger/ranger_test.go:964]: Then the default value `true` is used and some plans in this test cases will change.

File: distsql/request_builder.go (hunk 2 above)
Comment (qw4990) [reply distsql/request_builder.go:719]: Yep, I plan to remove this hard-code operation with the switch `tidb_regard_null_as_point` together a few months later if there is no new bug.

File: planner/core/plan_to_pb.go (hunk 3 above)
Comment (qw4990) [reply planner/core/plan_to_pb.go]: Updated.

File: util/ranger/ranger_test.go (hunk 1 above)
Comment (time-and-fate) [reply util/ranger/ranger_test.go:964]: This doesn't affect the correctness, right?

**Review by time-and-fate — APPROVED** [review]

File: planner/core/integration_test.go [hunk 4]
```diff
@@ -4835,6 +4835,50 @@ func (s *testIntegrationSerialSuite) TestRejectSortForMPP(c *C) {
...
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+
+	tk.MustExec("drop table if exists tpk")
+	tk.MustExec(`create table tuk (a int, b int, c int, unique key (a, b, c))`)
+	tk.MustExec(`create table tik (a int, b int, c int, key (a, b, c))`)
+	for _, va := range []string{"NULL", "1"} {
+		for _, vb := range []string{"NULL", "1"} {
+			for _, vc := range []string{"NULL", "1"} {
+				tk.MustExec(fmt.Sprintf(`insert into tuk values (%v, %v, %v)`, va, vb, vc))
```
Comment (xuyifangreeneyes) [inline planner/core/integration_test.go:4848]: Maybe it is better to insert each value twice expect (1,1,1). In this way we can check whether it returns multiple rows (expected)  or just a single row(wrong).

**Review by xuyifangreeneyes — APPROVED** [review]
LGTM

File: planner/core/integration_test.go (hunk 4 above)
Comment (qw4990) [reply planner/core/integration_test.go:4848]: updated

**Review by xuyifangreeneyes — APPROVED** [review]
LGTM

**Review by XuHuaiyu — APPROVED** [review]

**Pull request merged**